lib/vrl/stdlib/src/encode_logfmt.rs

```diff
@@ -0,0 +1,227 @@
+                result: Ok(r#"s'log_id=12345 lvl=info msg="This is a message"'"#),
+            },
+            Example {
+                title: "encode array",
+                source: r#"encode_logfmt([["lvl", "info"], ["msg", "This is a message"], ["log_id", 12345]])"#,
+                result: Ok(r#"s'lvl=info msg="This is a message" log_id=12345'"#),
+            },
+        ]
+    }
+}
+
+fn format_logfmt_string(f: &mut fmt::Formatter<'_>, str: &str) -> fmt::Result {
+    match str.find(' ') {
+        Some(_) => write!(f, r#""{}""#, str),
+        None => write!(f, "{}", str),
+    }
+}
+
+struct LogFmtFieldValueFormatter<'a> {
+    value: &'a Value,
+}
+
+impl std::fmt::Display for LogFmtFieldValueFormatter<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.value {
+            Value::Bytes(b) => {
+                let val = String::from_utf8_lossy(b);
+                format_logfmt_string(f, &val)
+            }
+            _ => {
+                let val = format!("{}", self.value);
+                format_logfmt_string(f, &val)
+            }
+        }
+    }
+}
+
+struct LogFmtFieldFormatter<'a> {
+    key: &'a str,
+    value: &'a Value,
+}
+
+impl std::fmt::Display for LogFmtFieldFormatter<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        format_logfmt_string(f, self.key)?;
+        let value_format = LogFmtFieldValueFormatter { value: self.value };
+        write!(f, "={}", value_format)
+    }
+}
+
+struct LogFmtFormatter {
+    value: Value,
+}
+
+impl std::fmt::Display for LogFmtFormatter {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.value {
+            Value::Object(map) => {
+                for (idx, (key, value)) in map.iter().enumerate() {
+                    let field_formatter = LogFmtFieldFormatter { key, value };
+                    if idx > 0 {
+                        f.write_str(" ")?;
+                    }
+                    write!(f, "{}", field_formatter)?;
+                }
+
+                Ok(())
+            }
+            Value::Array(arr) => {
+                for (idx, value) in arr.iter().enumerate() {
+                    if idx > 0 {
+                        f.write_str(" ")?;
+                    }
+
+                    match value {
+                        Value::Array(arr) if arr.len() == 2 => {
+                            let (key, value) = (&arr[0], &arr[1]);
+                            if let Value::Bytes(b) = key {
+                                let key_str = String::from_utf8_lossy(b);
+                                let field_formatter = LogFmtFieldFormatter {
+                                    key: &key_str,
+                                    value,
+                                };
+                                write!(f, "{}", field_formatter)?;
+                            }
+                        }
+                        _ => return Err(fmt::Error),
```

Comment:
Could we wrap these errors to provide more context to users? For example, I might like to see output like: `Failed to encode as logfmt, index 4 has more than two elements` and `Failed to encode as logfmt, index 4 is an integer, not an array`.

We've commonly just passed around strings for errors in the VRL stdlib so you can wrap them by `format!("Oh no, an error occurred: {}", err)`ing as they bubble up.